### PR TITLE
test(developer): check correct use of u16chr when second parameter could be null

### PR DESCRIPTION
**Pull request by markcsinclair**
Fix the use of u16chr() with *q==0 in u16tok() bug (and anywhere else it is similarly misused) - see issue [#11814](https://github.com/keymanapp/keyman/issues/11814)

Tests for `u16chr()` and `u16tok()` (both variants) are included in `gtest-kmxu16-test.cpp`

Partially fixes: [#11814](https://github.com/keymanapp/keyman/issues/11814) (another PR is needed for the first bug)

Related-to: [#11910](https://github.com/keymanapp/keyman/pull/11910)

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
I independently encountered the `u16tok` bug. I have pushed my branch so you can compare: #11910.

**Comment by markcsinclair** [comment]
I think the tests for u16chr and u16tok (both variants) are complete, so the next task is to identify and correct any additional incorrect uses of u16chr, where second parameter could be null.

**Comment by markcsinclair** [comment]
[PR #11910](https://github.com/keymanapp/keyman/pull/11910) has been merged into master, and in turn into this PR, so it is ready to be reviewed.

Compared to [PR #11910](https://github.com/keymanapp/keyman/pull/11910), this PR offers more extensive testing of `u16chr()` and the two variants of `u16tok()`, plus testing of two `Compiler.cpp` functions, `ProcessEthnologueStore()` and `LineTokenType()` which were (incorrectly) suspected of containing misuse of `u16chr()`.  The testing on `ProcessEthnologueStore()` did identify a different bug (see [#11955](https://github.com/keymanapp/keyman/issues/11955)) however.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 1]
```diff
@@ -556,7 +560,106 @@ TEST_F(CompilerTest, GetDelimitedString_test) {
...
+    u16cpy(str, u"c ");
+    p = str;
+    EXPECT_EQ(T_COMMENT, LineTokenType(&p));
+    EXPECT_EQ(0, p - str);
+
+    // comment without following space
+    u16cpy(str, u"c");
+    p = str;
+    EXPECT_EQ(T_UNKNOWN, LineTokenType(&p));
+    EXPECT_EQ(0, p - str);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:648]: Hmm, I would have thought that this should be treated as a comment. It turns out that `ReadLine()` strips the comment before it ever gets to `LineTokenType()`, so we don't get failure:

https://github.com/keymanapp/keyman/blob/bcec4461448f78498cac2b3711081546705b200a/developer/src/kmcmplib/src/Compiler.cpp#L3404-L3410

(In `ReadLine()` the input has a trailing `\n` so `iswspace(*(p+1))` is true.)

File: developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp [hunk 2]
```diff
@@ -1,60 +1,168 @@
...
+  strcpy(str, "abc<def>ghi");
+  const char *delim = "<>";
+	EXPECT_TRUE(!strcmp("abc", strtok(str, delim)));
+  EXPECT_TRUE(!strcmp("def", strtok(nullptr, delim)));
+  EXPECT_TRUE(!strcmp("ghi", strtok(nullptr, delim)));
+  EXPECT_EQ(nullptr, strtok(nullptr, delim));
+
+  // check sensitivity of ordering with multiple delimiters
+  strcpy(str, "abc<def>ghi");
+  const char *reverse_delim = "<>";
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp:163]: this doesn't match line 122

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looking good, two small items -- one of which should probably be treated as a latent (if not currently visible) bug in the compiler?

Noting that the title and description suggest changes to the compiler code, not just the tests -- is there something missing?

**Comment by markcsinclair** [comment]
> Noting that the title and description suggest changes to the compiler code, not just the tests -- is there something missing?

This PR originally contained a fix to the compiler whilst in draft, but then you developed a fix in PR [#11910](https://github.com/keymanapp/keyman/pull/11910).  I merged that into this PR, adopting your fix and extending my tests (see my comments above).  I will change the name of this PR to a reflect its final test-only content.

File: developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp (hunk 2 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-kmx_u16-test.cpp:163]: Good spot. Corrected.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 1 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:648]: I have left as is, except to add a comment to the test cf. `ReadLine()`.  Should I open an issue?

**Comment by mcdurdin** [comment]
> This PR originally contained a fix to the compiler whilst in draft, but then you developed a fix in PR [#11910](https://github.com/keymanapp/keyman/pull/11910). I merged that into this PR, adopting your fix and extending my tests (see my comments above). I will change the name of this PR to a reflect its final test-only content.

And ... I knew that and forgot. Sorry for the hassle!

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 1 above)
Comment (mcdurdin) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp]: Yeah, probably open an issue, add the low priority label to it.

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.80-alpha](https://keyman.com/downloads/releases/alpha/18.0.80)

**Pull request merged**